ranger/core/progress_tracker.py

```diff
@@ -0,0 +1,33 @@
+# This file is part of ranger, the console file manager.
+# License: GNU GPL version 3, see the file "AUTHORS" for details.
+
+"""
+A Progress Tracker of any Loadable that manages all the progress
+related statistics.
+"""
+
+import time
+
+class ProgressTracker(object):
+    def __init__(self, total_items=0):
+        self.total_items = total_items
+        self.done = 0
+        self.start_time = time.time()
+
+    def set_total_items(self, total_items):
+        self.total_items = total_items
+
+    def step(self, stepsize=1):
+        self.done += stepsize
+
+    @property
+    def percent(self):
+        if self.total_items == 0:
+            return 0
+        return (self.done / self.total_items) * 100
+
+    @property
+    def average_rate(self):
+        time_diff = time.time() - self.start_time
```

Comment:
Just something to think about: If the progress has been paused (for some reason), the average will be wrong. Furthermore, averaging over the entire period produces results that apply to the process as a whole, but for the most part these are time intervals that humans don't parse well or are even interested in. The current speed calculated with a sliding window, for example, is much more useful for speed calculation because it considers recency and values it more (or only) while abandoning old and outdated information.

To put it into actual suggestions: You could either create a reference point on every `step` call with both the progress at the time and the timestamp, or you create a fifo queue of limited size that you push discrete rates since the last update to and then calculate the average with those rates only.

Not saying you should definitely do that here or now, but something I wanted to mention nonetheless.

Reply:
Using FIFO buffer to implement sliding window speed calculation sounds good to me, added to my TODO list.